bftengine/src/bftengine/SigManager.cpp

```diff
@@ -304,10 +369,71 @@ void SigManager::setClientPublicKey(const std::string& key, PrincipalId id, KeyF
 bool SigManager::hasVerifier(PrincipalId pid) { return verifiers_.find(pid) != verifiers_.end(); }
 
 concord::crypto::SignatureAlgorithm SigManager::getMainKeyAlgorithm() const { return concord::crypto::EdDSA; }
-concord::crypto::ISigner& SigManager::getSigner() { return *mySigner_; }
+
+std::shared_ptr<EdDSAMultisigSigner> SigManager::getCurrentReplicaSigner() const {
+  ConcordAssert(ReplicaConfig::instance().singleSignatureScheme);
+  ConcordAssert(replicasInfo_.isIdOfReplica(myId_) || replicasInfo_.isRoReplica());
+  auto signer = CryptoManager::instance().getSigner(getReplicaLastExecutedSeq());
+  return signer;
+}
+
+std::shared_ptr<EdDSAMultisigSigner> SigManager::getLastReplicaSigner() const {
+  ConcordAssert(ReplicaConfig::instance().singleSignatureScheme);
+  ConcordAssert(replicasInfo_.isIdOfReplica(myId_) || replicasInfo_.isRoReplica());
+  auto latestSigner = CryptoManager::instance().getLatestSigners()[0];
+  return latestSigner;
+}
+
+std::pair<SeqNum, std::string> SigManager::getMyLatestPublicKey() const {
+  ConcordAssert(ReplicaConfig::instance().singleSignatureScheme);
+  ConcordAssert(replicasInfo_.isIdOfReplica(myId_));
+  auto [seq, latestVerifier] = CryptoManager::instance().getLatestVerifiers()[0];
+  auto pubKey = latestVerifier->getVerifier(myId_).getPubKey();
+  LOG_DEBUG(GL, KVLOG(pubKey));
+  return {seq, pubKey};
+}
+
+std::string SigManager::getSelfPrivKey() const {
+  if (ReplicaConfig::instance().singleSignatureScheme) {
+    return getCurrentReplicaSigner()->getPrivKey();
+  }
+  return mySigner_->getPrivKey();
+}
+
+std::array<std::string, replicaIdentityHistoryCount> SigManager::getPublicKeyOfVerifier(uint32_t id) const {
+  std::array<std::string, replicaIdentityHistoryCount> publicKeys;
+
+  if (ReplicaConfig::instance().singleSignatureScheme && replicasInfo_.isIdOfReplica(id)) {
+    int i = 0;
+    for (auto [seq, multisigVerifier] : CryptoManager::instance().getLatestVerifiers()) {
+      UNUSED(seq);
+      if (multisigVerifier.get() == nullptr) {
+        continue;
+      }
+      auto& verifier = multisigVerifier->getVerifier(id);
+      publicKeys[i] = verifier.getPubKey();
+      ++i;
+    }
+  } else if (verifiers_.count(id)) {
+    publicKeys[0] = verifiers_.at(id)->getPubKey();
+  }
+
+  return publicKeys;
+}
+
 const concord::crypto::IVerifier& SigManager::getVerifier(PrincipalId otherPrincipal) const {
   return *verifiers_.at(otherPrincipal);
 }
 
+void SigManager::setReplicaLastExecutedSeq(SeqNum seq) {
```

Comment:
Do you use this method in the RO replica?
is a ROR even capable of exchanging its keys?